feat(adr): initial ADR to address service list (#377)
edgexfoundry/edgex-docs: docs_src/design/adr/0019-Service-List.md

- The CLI could refer to a dynamic list of which services are operational and which device services and application services to include in start/stop operations.  *Note: Alternatively, the CLI could simply issue a command to a service and if the service is not available, issue an error response.  However, if a new service was added (such as a new device service), the CLI would not know about it unless it had access to some dynamic list of services.*
- Security proxy setup requires a list of services for which a reverse proxy route should be established.  This is a subset of the full list of EdgeX services.
- Core Metadata must maintain a list of device-type services.  This list is used to validate device additions (i.e., any new device definition must include a valid named device service which it is going to be associated with).  This list is also used when routing commands (via core command) to the appropriate device service.

In the future, there may be other internal or 3rd party needs to what services comprise an EdgeX deployment.

### Current Situation

Currently, the list of services is haphazardly managed by the needing service.

> - The security bootstrapping services use a combination of staic lists for known EdgeX services and environment variables for services that are not statically known.  (This list is needed for both Vault and Consul to setup appropriate access control lists.)

Reviewer: `staic` -> `static`

- In the case that the services are distributed (especially when device services run on different hosts - as is likely), access to a distributed file system or volume mount becomes more of an issue and requires additional technology.  *Note: distributing device services while in secure mode is now more complex due to the fact that all services require Vault access and the services get their Vault token from file.  Distributed device services in secure mode would therefore would require a distributed file system.*
- The service list would have to be a list of services and service locations (URI).  Therefore, the service list will really be a map of service names (keys) and the location of the services URIs (values) and potentially other information such as `known secrets` as used by the secret store setup (`redisdb` is the only `known secret` to date).  This will help support things like the API gateway setup as well.  In some cases, just a list of the service names is needed (in security token setup for example).  In this latter case, the service would just pick off the service keys.

> - Generation of security tokens for the new services under most of the conceived alternatives is complex.  In the current security bootstrapping, secret store setup will require a restart of all the services in order to add or remove a service.  The ORRA project and some adopters are clear that they would like to have the ability to add/remove services dynamically at runtime.  *Note: in some circumstances, services can be added/removed dynamically today, but not when EdgeX security services are enabled.  CLI and UI also do not allow for the addition/remove of services - such as device or application services.*

Reviewer 2: This is now out of date with recent enhancements. Secret Store setup can now be restarted w/o requiring the other service to be restarted.

# EdgeX Service List

## Status
**Draft as of April 25, 2021**
Proposed for Jakarta or later release.

## Context

Several EdgeX Foundry services today require a list of the services that are part of an EdgeX deployment.  That is to say, that several EdgeX services need to know the names and locations of other EdgeX services that are or will be part of the EdgeX.

The current list of needs include:

> - Security bootstrapping service requires a list of services in order to know which services need a file token for Vault access and to add shared secrets to Vault.

Reviewer 3: ```suggestion
- Security bootstrapping service requires a list of services in order to know which services need a file token for Vault access and to add shared secrets to Vault.
- Consul bootstrapping service requires a list of services in order to generate an appropriate per-service Consul ACL as well as to configure Vault to allow the service to request a Consul token with that ACL.
```

- System management agent service requires a list of services in order to know which services it can provide status, configuration and metrics for client requestors.  The SMA also needs to know which services to include in any start/stop/restart operations.
- The user interface needs to know which services, specifically device and application services, are available in order to provide interaction with those services, and potentially to display some metrics on operational services.
- The CLI needs to know which services are operational (for some interaction) and which device services and application services to include in start/stop operations.

> 

Reviewer 3: ```suggestion
- Security proxy setup requires a lot of services for which a reverse proxy route should be established.  This is a subset of the full list of EdgeX services.
```

> 

Reviewer 4: core-metadata needs (and currently maintains) a list of device-type services. This is used to validate device additions (the device definition must include a valid named device service to which it is associated). It is also needed when routing device commands to the appropriate service.

(Should the existing facility in core-metadata be expanded to cover registration of all services?)

Author: good call out Iain. Added this consideration.

In the future, there may be other internal or 3rd party needs to what services comprise an EdgeX deployment.

### Current Situation

Currently, the list of services is haphazardly managed by the needing service.

> - The security bootstrapping services are being built to use an environment variable to provide it with a list of services at startup.

Reviewer 3: ```suggestion
- The security bootstrapping services are being built to use an environment variable to provide it with a list of services at startup.  (This list is needed for both Vault and Consul to setup appropriate access control lists.)
```

- The user interface and CLI do not deal with service lists today and assume all services are operational (with known service host and port defined in configuration).
- The SMA has a static/hardcoded list built into the codebase.  Clearly the least desirable means of dealing with the list of services as this not only requires a restart of the service, but also code changes and redeployment.

### Requirements

1. It is desired that the service list be addressed in a consistent manner.  It is further desired that there be one single, authoritative source of the service list information for all EdgeX services and 3rd party requestors.  While the source of the service list may change under certain runtime circumstances (for example – whether operating with or without security), once EdgeX is running the authoritative source remains the same for all services.

> 2. Any changes to the list of services must adhere to ACID transactional properties.  That is, any change to the list of services must be atomic, consistent, isolated and durable with regard to any operations on the service list.  In other words, no two services should ever be able to get a different list of services when making a request of EdgeX for the list of services.

Reviewer 3: These two statements aren't the same.  "no two services should ever be able to get a different list of services when making a request of EdgeX for the list of services." means that the list is static. 

ACID-compliance means:
(A) Suppose there is an interface to add or delete a list of services.  Atomicity guarantees that no client sees a partially-executed list of adds or removes.
(C) Suppose there was a rule that removing a service from the list also removed it from the reverse proxy and that there is a 1:1 mapping between the two.  Consistency says that no client will ever observe a proxy rule without a accompanying service or a service without an accompanying proxy rule.  Both must be observed at the same time.
(I) If multiple concurrent updates to the service list are possible, a client will observe them as if they were applied in some sequential order.
(D) Means if the service list is modified, if the component in charge of the service list crashes and is restarted, the service list should be restored to its last modified state.

ACID properties are meaningless if the list of services is static.

Author: Bryon - what I meant by this (and have changed the language of this section to hopefully clarify) is that like in database transactions (that adhere to ACID), if service A makes a change to the the service list and then service B requests the service list, that service B should see the change made by service A with its request.  The service list is not static (or at least at this point in discussion we are thinking it will not be static).

> 3. EdgeX services may be added or removed (either intentionally or unintentionally due to failures) all the time.  The list of services should reflect, as best it can, the current list of EdgeX services – that is be dynamic and not a static list at the start/bootstrap of the EdgeX instance.  In particular, new device services and application services are likely to be added to the EdgeX instance after the system is started.  The list should eventually reflect the additions.

Reviewer 3: The phrases "reflect, as best it can" and "should eventually reflect" are consistency violations, as defined in a previous comment.

Author: agree this made it seem non-ACID and subject to eventual consistency.  I took out this language.

> 

Reviewer 3: Additions and deletions from the EdgeX service list need to be authenticated and authorized.

Author: Yes - added bullet.

> 4. While Consul is often used as the configuration/registry service, EdgeX was constructed to operate off of local configuration in the absence of the central config/registry service (for development purposes or in order to reduce the resource needs of the platform).  The service list need should be provided when a central configuration/registry service like Consul is provided and even when services operate off local configuration.  Per #1 above, all services should get the service list information in the same way, but that way may change depending on the existence of a central configuration/registration service.

Reviewer 3: ```suggestion
4. While Consul is often used as the configuration/registry service, EdgeX was constructed to operate off of local configuration in the absence of the central config/registry service (for development purposes or in order to reduce the resource needs of the platform).  The service list need should be provided when a central configuration/registry service like Consul is provided and even when services operate off local configuration.  Per #1 above, all services should get the service list information in the same way, but that may change depending on the existence of a central configuration/registration service.
```

- In some cases, a service may be “disabled” at runtime because of resource constraints (example: notification service is turned off as it is not used and there is a need to operate with limited memory and CPU capability).  Security would still need to know about disabled services in order to create a token for such services that may be enabled later.  However, the other services (or 3rd party applications) may need to know that they are disabled vs enabled.
- In the case that the services are distributed (especially when device services run on different hosts - as is likely), access to a distributed file system or volume mount becomes more of an issue and requires additional technology.
- The service list would have to be a list of services and service locations (URI).  Therefore, the service list will really be a map of service names (keys) and the location of the services URIs (values).  This will help support things like the API gateway setup as well.  In some cases, just a list of the service names is needed (in security token setup for example).  In this latter case, the service would just pick off the service keys.

> - Generation of security tokens for the new services under most of the conceived alternatives is complex.  In the current security bootstrapping, secret store setup will require a restart of all the services in order to add or remove a service.  The ORRA project and some adopters are clear that they would like to have the ability to add/remove services dynamically at runtime.

Reviewer 3: > The ORRA project and some adopters are clear that they would like to have the ability to add/remove services dynamically at runtime.

Should clarify that this can be done already, but not when EdgeX security services are enabled.

Author: Bryon - I don't think this is entirely true.  Adding or removing new services (dynamically) would not work for SMA, CLI or UI today.  It might be easier to address these changes without security services, but would still require some change.

> - From a security perspective, being able to add/remove new services dynamically opens up a rather large attack vector and vulnerability.

Reviewer 3: ```suggestion
- From a security perspective, being able to add/remove new services dynamically opens up a rather large attack surface.
```

Don't have vulnerabilities at this point since the implementation is yet to be defined.

### Implementation Alternative Discussion

- A service list could be made available in a configuration file in the file system (non-Docker) or volume mount (in a Docker situation).
  - All services would get the configuration in a consistent way (#1)

>   - It would also mean that there is a single, authoritative list for all services and it could be made to be ACID safe (#2 and #3).

Reviewer 3: > It would also mean that there is a single, authoritative list for all services and it could be made to be ACID safe (#2 and #3).
I disagree that concurrent updates to a configuration file could be made ACID compliant.

Author: Point taken.  There are technologies such as Transactional NTFS from Microsoft that make some claims with regard to transactional activity on files, but finding an easy to implement, ACID transactional, file based solution to the service list issue would be a challenge.  I'll note this.

>   - The file would have to be validated whenever edited in order not to bring down all the services with bad contents.

Reviewer 3: > The file would have to be validated whenever edited in order not to bring down all the services with bad contents.

Integrity of said configuration file cannot be protected if it is allowed to be edited by hand.

Author: Noted

  - This alternative would solve the requirement that the solution operate with or without Consul (#4).  UI and CLI should have access to the same file or volume mount.
  - There is a question about who or what service would put this file into the file system or volume mount – especially since security bootstrapping would need it first but this file would be needed in non-secure runtime as well.

>   - There isn’t a central “dummy” service that would own the file – especially for dev/test.

Reviewer 5: We may use "placeholder" instead of "dummy"
https://github.com/ietf/terminology

Author: Just removed the word dummy.  I don't think we even need the term placeholder here.  The point is we don't have a service that would be the owner of this file/info.

>   - It has to be a service that is always there (like core metadata), but also available from the start of all services (which metadata is not in the case of security services).

Reviewer 5: Please mention the service should be run with non-root user

  - Services would have to monitor the file or volume for changes in order to be able to address the add/removal of services at runtime.
  - How would the security bootstrapping address the add/removal of services and get new tokens established?  The bootstrapping service isn’t even long running to do this.
- Have global configuration in Consul (or config/reg service) that everyone reads.
  - When Consul is not used, have a common service list in each service configuration.toml file that everyone reads.

>   - While the means to get the configuration is consistent in this solution alternative (#1), there is not a single, authoritative list of services when using local configuration and making ACID transactions (#3) across the multiple files would be difficult – especially if distributed across hosts.

Reviewer 3: This bullet item is confusing.  This section is supposed to be talking about "Have global configuration in Consul (or config/reg service) that everyone reads."  Also, aren't EdgeX's configuration.toml considered read-only?

Author: tried to clarify and clean this up.  Bottom line, Consul could be used to make things consistent and ACID safe with regard to service list, but since people can use configuration files in place of Consul.  When each service has its own config file (and doesn't use Consul), consistency and any ideas of ACID transactions against the multiple configuration files goes bye bye.

>   - Using Consul would make sure that the service list is valid and that any change operation is ACID.

Reviewer 3: > service list is valid

What constitutes a validity when talking about a service list?  XML, for example, be both valid and well-formed.  Well-formed-ness means that it can be read by an XML parser.  Valid means that it confirms to a known schema.  You can make XML that is well-formed but not valid.

> any change operation is ACID

Depends on how a transaction is defined, for example, does the API gateway need to make a new route when adding a service?

# EdgeX Service List

## Status
**Draft as of April 25, 2021**
Proposed for Jakarta or later release.

## Context

Several EdgeX Foundry services today require a list of the services that are part of an EdgeX deployment.  That is to say, that several EdgeX services need to know the names and locations of other EdgeX services that are or will be part of the EdgeX.

The current list of needs include:

- Security bootstrapping service requires a list of services in order to know which services need a file token for Vault access and to add shared secrets to Vault.

> - Consul bootstrapping service requires a list of services in order to generate an appropriate per-service Consul ACL as well as to configure Vault to allow the service to request a Consul token with that ACL.

Author: @bnevis-i added the second to my first.  Need @bnevis-i input and wording on how to combine these.

> - Consul bootstrapping service requires a list of services in order to generate an appropriate per-service Consul ACL as well as to configure Vault to allow the service to request a Consul token with that ACL.

Reviewer 3: "The security-secretstore-setup service and the Consul bootstrapping script are closely intertwined as a services' access to Consul is granted via the EdgeX's secret store in secure mode. Each EdgeX service owns a slide of the configuration store as well as the secret store into which shared secrets such as Redis credentials are injected. The list of services directly drives the sub-partitioning of these data stores, enables security linkage between Vault and Consul, and controls generation of a file-based token that grants' services the ability to connect to the secret store and consequently Consul."?

- System management agent service requires a list of services in order to know which services it can provide status, configuration and metrics for client requestors.  The SMA also needs to know which services to include in any start/stop/restart operations.
- The user interface needs to know which services, specifically device and application services, are available in order to provide interaction with those services, and potentially to display some metrics on operational services.

> - The CLI needs to know which services are operational (for some interaction) and which device services and application services to include in start/stop operations.

Author: Noted the optional nature of this need and how the CLI could just issue the command and error out.  What this does not address however is that as new services are added (such as a new device service), the CLI would not know about them without a dynamic list reference.

> - Security proxy setup requires a lot of services for which a reverse proxy route should be established.  This is a subset of the full list of EdgeX services.

Reviewer: Should "a lot of services" be "a list of services" ?

- Core Metadata must maintain a list of device-type services.  This list is used to validate device additions (i.e., any new device definition must include a valid named device service which it is going to be associated with).  This list is also used when routing commands (via core command) to the appropriate device service.

In the future, there may be other internal or 3rd party needs to what services comprise an EdgeX deployment.

### Current Situation

Currently, the list of services is haphazardly managed by the needing service.

> - The security bootstrapping services are being built to use an environment variable to provide it with a list of services at startup.  (This list is needed for both Vault and Consul to setup appropriate access control lists.)

Reviewer 3: ```suggestion
- The security bootstrapping services use a combination of static lists for known EdgeX services and environment variables for services that are not statically known.  (This list is needed for both Vault and Consul to setup appropriate access control lists.)
```

- The user interface and CLI do not deal with service lists today and assume all services are operational (with known service host and port defined in configuration).
- The SMA has a static/hardcoded list built into the codebase.  Clearly the least desirable means of dealing with the list of services as this not only requires a restart of the service, but also code changes and redeployment.

### Requirements

1. It is desired that the service list be addressed in a consistent manner.  It is further desired that there be one single, authoritative source of the service list information for all EdgeX services and 3rd party requestors.  While the source of the service list may change under certain runtime circumstances (for example – whether operating with or without security), once EdgeX is running the authoritative source remains the same for all services.

> 2. Any changes to the list of services must adhere to ACID transactional properties.  That is, any change to the list of services must be atomic, consistent, isolated and durable with regard to any operations on the service list.  In other words, if a service A makes a change to the list of services, service B should see that change and now have the same list as service A (once the change has been made in some sort of transactional boundry).  No two services should ever be able to get a different list of services when making a request of EdgeX for the list of services.

Author: Added the important clause of "at the same time"

3. EdgeX services may be added or removed (either intentionally or unintentionally due to failures) all the time.  The list of services should reflect the current list of EdgeX services – that is be dynamic and not a static list at the start/bootstrap of the EdgeX instance.  In particular, new device services and application services are likely to be added to the EdgeX instance after the system is started.

4. Additions and deletions from the EdgeX services to the service list need to be authenticated and authorized.

> 5. While Consul is often used as the configuration/registry service, EdgeX was constructed to operate off of local configuration in the absence of the central config/registry service (for development purposes or in order to reduce the resource needs of the platform).  The service list need should be provided when a central configuration/registry service like Consul is provided and even when services operate off local configuration.  Per #1 above, all services should get the service list information in the same way, but that may change depending on the existence of a central configuration/registration service.

Reviewer 2: remove `configuration/` from `The service list need should be provided when a central configuration/registry service like Consul` since it is the registry that is needed in this case, not config provider.

### Considerations

- In some cases, the list of services is needed before the services are up (security needs to know so appropriate tokens can be generated and handed out when the services bootstrap).
- In some cases, a service may be “disabled” at runtime because of resource constraints (example: notification service is turned off as it is not used and there is a need to operate with limited memory and CPU capability).  Security would still need to know about disabled services in order to create a token for such services that may be enabled later.  However, the other services (or 3rd party applications) may need to know that they are disabled vs enabled.

> - In the case that the services are distributed (especially when device services run on different hosts - as is likely), access to a distributed file system or volume mount becomes more of an issue and requires additional technology.

Reviewer 2: Note that distributed devices services in secure mode has been made more complex due to all services require Vault access and they get their Vault token from file. For this to work it would require a distributed file system.

Reviewer 3: > Note that distributed devices services in secure mode has been made more complex due to all services require Vault access and they get their Vault token from file. For this to work it would require a distributed file system.

FYI: SPIFFE/SPIRE tooling also supports distribution.

> - The service list would have to be a list of services and service locations (URI).  Therefore, the service list will really be a map of service names (keys) and the location of the services URIs (values).  This will help support things like the API gateway setup as well.  In some cases, just a list of the service names is needed (in security token setup for example).  In this latter case, the service would just pick off the service keys.

Reviewer 2: For Secret Store Setup, we also have a list of services that need their Secret Store seeded with specified `known secrets`. Current, `redisdb` is the only `known secret`.

Author: added

  - The file would have to be validated whenever edited in order not to bring down all the services with bad contents.  Integrity of the configuration file could not be protected if someone edited the file by hand.
  - This alternative would solve the requirement that the solution operate with or without Consul (#4).  UI and CLI should have access to the same file or volume mount.
  - There is a question about who or what service would put this file into the file system or volume mount – especially since security bootstrapping would need it first but this file would be needed in non-secure runtime as well.
  - There isn’t a central service that would own the file – especially for dev/test.
  - It has to be a service that is always there (like core metadata), but also available from the start of all services (which metadata is not in the case of security services).  Ideally, this service should not require root user privileaes.
  - Services would have to monitor the file or volume for changes in order to be able to address the add/removal of services at runtime.

>   - How would the security bootstrapping address the add/removal of services and get new tokens established?  The bootstrapping service isn’t even long running to do this.

Reviewer 3: This isn't really such a big deal.  There is specific configuration `TokenProviderType = "oneshot"` to code this in.  It is entirely possible to extend this to create, for example, a "forking" token provider that would run in the background indefinitely.

Reviewer 3: Some rework, however, would be needed for Consul, as we are doing that in the pre-launch script, which really does run only once.

  - Using Consul would make sure that the service list is well formed (meaining adherence to format, but this does not mean that the service list necessarily adheres to any validity rules unless additional code interacting with Consul changes is checking for well formed-ness and validity) and that any change operation is ACID (but only with regard to Consul and the service list - this does not mean the transaction would incorporate additional needs such as the API Gateway updating new routes).
  - When using the local configuration, some of the same issues exist as listed in the first alternative (validation of the file, etc.).
  - Services would have to be notified by Consul if the service list changes or monitor the local config for a service list change in order to be able to address the add/removal of services at runtime.
  - Again, how would the security bootstrapping address the add/removal of services and get new tokens established?  The bootstrapping service isn’t even long running to do this.

>   - We could use Consul for dynamic service needs and when Consul is not used resolve that the service list is static and requires a restart of the services.

Reviewer 2: How about a Registry implementation that uses new API on Core Metadata for simple registry. Yes, now Core Metadata has to be up first...

Reviewer 4: Could also add a key-value store API to core-metadata, then all of the "but what if we haven't got a registry / config provider?" code goes away :)

Author: added as an alternate

>   - There is a mutually inclusive event issue (aka chicken and egg problem) with this solution in that we have to secure Consul via the security bootstrapping providing it a token and a list of all the services, but Consul is the place where the security service would get the list of all services.  The circumstances prohibit both to be true.

Reviewer 3: "and a list of all the services".  This might be boiling the ocean a bit.  All that is needed is an induction, that is, an ability to dynamically onboard the first service.  Repeat until all services are onboarded.

- Use environment variables to provide the list of services.

>   - Each micro service would use the same env var to get the list of services.

Reviewer 6: Right, but you can't really do this. They can use the same env var name, but the actually value is a per-process thing, unless you use an OS/orchestration mechanism to ensure that the value is the same for all processes...

  - Use of an env var is a bit cumbersome in setup of all services – especially when the env var changes.  This is considered “unmanageable”.
  - Adding or removing a service at runtime get tricky under this alternative (you can’t change the env var after startup)
  - This would be a single, authoritative location for service list information.
  - ACID needs would be moot as the env var would not be changed post startup.

## Decision

To be determined as part of a TSC and monthly architect’s meeting.

## Consequences

A key factor in determining a solution is whether EdgeX will support dynamic addition of services.  If the list of services is static and requires a restart of all of EdgeX when the list changes, the solutions are many and implementation straightforward.

Adopters / users are requesting the ability to add/remove services at runtime (and operate in secure mode).

> ## References

Author: Per request from @tonyespy, I will add a reference to the registry ADR in this section on the next cut.  https://docs.edgexfoundry.org/2.0/design/adr/0018-Service-Registry/
